src/crl.rs

```diff
@@ -603,6 +623,122 @@ impl TryFrom<u8> for RevocationReason {
+        const INDIRECT_CRL_TAG: u8 = CONTEXT_SPECIFIC | 4;
+        const ONLY_CONTAINS_ATTRIBUTE_CERTS_TAG: u8 = CONTEXT_SPECIFIC | 5;
+
+        let mut result = IssuingDistributionPoint {
+            distribution_point: None,
+            only_contains_user_certs: false,
+            only_contains_ca_certs: false,
+            only_some_reasons: None,
+            indirect_crl: false,
+            only_contains_attribute_certs: false,
+        };
+
+        // Note: we can't use der::optional_boolean here because the distribution point
+        //       booleans are context specific primitives and der::optional_boolean expects
+        //       to unwrap a Tag::Boolean constructed value.
+        fn decode_bool(value: untrusted::Input) -> Result<bool, Error> {
+            let mut reader = untrusted::Reader::new(value);
+            let value = reader.read_byte()?;
+            if !reader.at_end() {
+                return Err(Error::BadDer);
+            }
+            match value {
+                0xFF => Ok(true),
+                0x00 => Ok(false), // non-conformant explicit encoding allowed for compat.
+                _ => Err(Error::BadDer),
+            }
+        }
+
+        // RFC 5280 section §4.2.1.13:
+        let mut der = untrusted::Reader::new(der);
+        der::nested(&mut der, der::Tag::Sequence, Error::BadDer, |der| {
+            while !der.at_end() {
+                let (tag, value) = der::read_tag_and_get_value(der)?;
+                match tag {
+                    DISTRIBUTION_POINT_TAG => {
+                        set_extension_once(&mut result.distribution_point, || Ok(value))?
+                    }
+                    ONLY_CONTAINS_USER_CERTS_TAG => {
+                        result.only_contains_user_certs = decode_bool(value)?
+                    }
+                    ONLY_CONTAINS_CA_CERTS_TAG => {
+                        result.only_contains_ca_certs = decode_bool(value)?
+                    }
+                    ONLY_CONTAINS_SOME_REASONS_TAG => {
+                        set_extension_once(&mut result.only_some_reasons, || {
+                            der::bit_string_flags(value)
+                        })?
+                    }
+                    INDIRECT_CRL_TAG => result.indirect_crl = decode_bool(value)?,
+                    ONLY_CONTAINS_ATTRIBUTE_CERTS_TAG => {
+                        result.only_contains_attribute_certs = decode_bool(value)?
+                    }
+                    _ => return Err(Error::BadDer),
+                }
+            }
+
+            Ok(result)
+        })
+    }
+
+    /// Return the distribution point names (if any).
+    #[allow(dead_code)] // TODO(@cpu): remove this once used in CRL validation.
+    pub(crate) fn names(&self) -> Result<Option<DistributionPointName<'a>>, Error> {
+        self.distribution_point
+            .map(DistributionPointName::from_der)
+            .transpose()
+    }
+
+    fn validate(&self) -> Result<(), Error> {
```

Comment:
I'm generally a little suspicious of separate `validate()` methods. Why not make this part of `from_der()`, per the "Parse, don't validate" idea of only succeeding in construction if the state is actually valid?